File: MVPJavaRx/app/src/main/java/com/vascome/fogtail/ui/base/fragments/BaseFragment.java
package com.vascome.fogtail.ui.base.fragments;

import android.os.Handler;
import android.os.Looper;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;

import com.vascome.fogtail.FogtailApplication;
import com.vascome.fogtail.appmodules.ApplicationModule;

import javax.inject.Inject;
import javax.inject.Named;

/**
 * Created by vasilypopov on 11/23/17
 * Copyright (c) 2017 MVPJava. All rights reserved.
 */

public abstract class BaseFragment extends Fragment {

    @Inject
    @Named(ApplicationModule.MAIN_THREAD_HANDLER)
    Handler mainThreadHandler;

    protected void runOnUiThreadIfFragmentAlive(@NonNull Runnable runnable) {
        if (Looper.myLooper() == Looper.getMainLooper() && isFragmentAlive()) {
            runnable.run();
        } else {
            mainThreadHandler.post(() -> {
                if (isFragmentAlive()) {
                    runnable.run();
                }
            });
        }
    }

    private boolean isFragmentAlive() {
        return getActivity() != null && isAdded() && !isDetached() && getView() != null && !isRemoving();
    }

    @Override
    public void onDestroy() {
        FogtailApplication.get(getContext()).appComponent().leakCanaryProxy().watch(this);
        super.onDestroy();
    }
}

File: MosbyMVPKotlin/app/src/main/java/com/vascome/fogtail/presentation/base/fragments/BaseFragment.kt
package com.vascome.fogtail.presentation.base.fragments

import android.os.Bundle
import com.hannesdorfmann.mosby3.mvp.MvpPresenter
import com.hannesdorfmann.mosby3.mvp.MvpView
import com.hannesdorfmann.mosby3.mvp.viewstate.MvpViewStateFragment
import com.hannesdorfmann.mosby3.mvp.viewstate.ViewState
import com.vascome.fogtail.di.Injectable
import com.vascome.fogtail.utils.LeakCanaryProxy

import javax.inject.Inject


/**
 * Created by vasilypopov on 11/23/17
 * Copyright (c) 2017 MVPJava. All rights reserved.
 */

abstract class BaseFragment<V : MvpView, P : MvpPresenter<V>, VS: ViewState<V>> : MvpViewStateFragment<V, P, VS>(), Injectable {

    @Inject
    lateinit var leakCanaryProxy: LeakCanaryProxy

    protected fun runOnUI(runnable: Runnable) {
        runnable.run()
    }
    override fun onDestroy() {
        leakCanaryProxy.watch(this)
        super.onDestroy()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        //AndroidSupportInjection.inject(this)
        super.onCreate(savedInstanceState)
        retainInstance = true
    }
}
